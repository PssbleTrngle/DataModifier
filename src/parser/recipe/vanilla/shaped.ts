import RecipeParser, { Recipe, replace } from '..'
import { Ingredient, Predicate, Result } from '../../../common/ingredient'
import { RecipeDefinition } from '../../../schema/recipe'
import { mapValues } from 'lodash-es'

export type ShapedRecipeDefinition = RecipeDefinition &
   Readonly<{
      key: Record<string, Ingredient>
      pattern: string[]
      result: Result
   }>

class ShapedRecipe extends Recipe<ShapedRecipeDefinition> {
   getIngredients(): Ingredient[] {
      return Object.values(this.definition.key)
   }

   getResults(): Result[] {
      return [this.definition.result]
   }

   replaceIngredient(from: Predicate<Ingredient>, to: Ingredient): Recipe {
      return new ShapedRecipe({
         ...this.definition,
         key: mapValues(this.definition.key, replace(from, to)),
      })
   }

   replaceResult(from: Predicate<Ingredient>, to: Result): Recipe {
      return new ShapedRecipe({
         ...this.definition,
         result: to,
      })
   }
}

export default class ShapedParser extends RecipeParser<ShapedRecipeDefinition, ShapedRecipe> {
   create(definition: ShapedRecipeDefinition): ShapedRecipe {
      return new ShapedRecipe(definition)
   }
}
