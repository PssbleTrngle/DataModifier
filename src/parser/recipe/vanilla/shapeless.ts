import RecipeParser, { Recipe, replace } from '..'
import { Ingredient, Predicate, Result } from '../../../common/ingredient'
import { RecipeDefinition } from '../../../schema/recipe'

type ShapelessRecipeDefinition = RecipeDefinition &
   Readonly<{
      ingredients: Ingredient[]
      result: Result
   }>

class ShapelessRecipe extends Recipe<ShapelessRecipeDefinition> {
   getIngredients(): Ingredient[] {
      return this.definition.ingredients
   }

   getResults(): Result[] {
      return [this.definition.result]
   }

   replaceIngredient(from: Predicate<Ingredient>, to: Ingredient): Recipe {
      return new ShapelessRecipe({
         ...this.definition,
         ingredients: this.definition.ingredients.map(replace(from, to)),
      })
   }

   replaceResult(from: Predicate<Ingredient>, to: Result): Recipe {
      return new ShapelessRecipe({
         ...this.definition,
         result: to,
      })
   }
}

export default class ShapelessParser extends RecipeParser<ShapelessRecipeDefinition, ShapelessRecipe> {
   create(definition: ShapelessRecipeDefinition): ShapelessRecipe {
      return new ShapelessRecipe(definition)
   }
}
