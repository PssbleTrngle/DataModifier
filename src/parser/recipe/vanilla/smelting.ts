import RecipeParser, { Recipe } from '..'
import { Ingredient, Predicate, Result } from '../../../common/ingredient'
import { RecipeDefinition } from '../../../schema/recipe'

type SmeltingRecipeDefinition = RecipeDefinition &
   Readonly<{
      ingredient: Ingredient
      result: Result
      experience?: number
      cookingTime?: number
   }>

class SmeltingRecipe extends Recipe<SmeltingRecipeDefinition> {
   getIngredients(): Ingredient[] {
      return [this.definition.ingredient]
   }

   getResults(): Result[] {
      return [this.definition.result]
   }

   replaceIngredient(from: Predicate<Ingredient>, to: Ingredient): Recipe {
      return new SmeltingRecipe({
         ...this.definition,
         ingredient: to,
      })
   }

   replaceResult(from: Predicate<Ingredient>, to: Result): Recipe {
      return new SmeltingRecipe({
         ...this.definition,
         result: to,
      })
   }
}

export default class SmeltingParser extends RecipeParser<SmeltingRecipeDefinition, SmeltingRecipe> {
   create(definition: SmeltingRecipeDefinition): SmeltingRecipe {
      return new SmeltingRecipe(definition)
   }
}
