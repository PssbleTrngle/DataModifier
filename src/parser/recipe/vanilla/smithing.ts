import RecipeParser, { Recipe, replace } from '..'
import { Ingredient, Predicate, Result } from '../../../common/ingredient'
import { RecipeDefinition } from '../../../schema/recipe'

type SmithingRecipeDefinition = RecipeDefinition &
   Readonly<{
      base: Ingredient
      addition: Ingredient
      result: Result
   }>

class SmithingRecipe extends Recipe<SmithingRecipeDefinition> {
   getIngredients(): Ingredient[] {
      return [this.definition.base, this.definition.addition]
   }

   getResults(): Result[] {
      return [this.definition.result]
   }

   replaceIngredient(from: Predicate<Ingredient>, to: Ingredient): Recipe {
      return new SmithingRecipe({
         ...this.definition,
         base: replace(from, to)(this.definition.base),
         addition: replace(from, to)(this.definition.addition),
      })
   }

   replaceResult(from: Predicate<Ingredient>, to: Result): Recipe {
      return new SmithingRecipe({
         ...this.definition,
         result: to,
      })
   }
}

export default class SmithingParser extends RecipeParser<SmithingRecipeDefinition, SmithingRecipe> {
   create(definition: SmithingRecipeDefinition): SmithingRecipe {
      return new SmithingRecipe(definition)
   }
}
