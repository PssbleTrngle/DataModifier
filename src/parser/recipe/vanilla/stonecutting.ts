import RecipeParser, { Recipe } from '..'
import { Ingredient, Predicate, Result } from '../../../common/ingredient'
import { RecipeDefinition } from '../../../schema/recipe'

type StonecuttingRecipeDefinition = RecipeDefinition &
   Readonly<{
      ingredient: Ingredient
      result: string
      count?: number
   }>

class StonecuttingRecipe extends Recipe<StonecuttingRecipeDefinition> {
   getIngredients(): Ingredient[] {
      return [this.definition.ingredient]
   }

   getResults(): Result[] {
      return [{ item: this.definition.result, count: this.definition.count }]
   }

   replaceIngredient(from: Predicate<Ingredient>, to: Ingredient): Recipe {
      return new StonecuttingRecipe({
         ...this.definition,
         ingredient: to,
      })
   }

   replaceResult(from: Predicate<Ingredient>, to: Result): Recipe {
      if (!('item' in to)) throw new Error('stonecutting does only support item results')

      return new StonecuttingRecipe({
         ...this.definition,
         result: to.item,
         count: to.count ?? 1,
      })
   }
}

export default class StonecuttingParser extends RecipeParser<StonecuttingRecipeDefinition, StonecuttingRecipe> {
   create(definition: StonecuttingRecipeDefinition): StonecuttingRecipe {
      return new StonecuttingRecipe(definition)
   }
}
