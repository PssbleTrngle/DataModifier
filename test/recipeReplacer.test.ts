import RecipeEmitter from '../src/emit/recipe'
import PackLoader from '../src/loader/pack'
import { ShapedRecipeDefinition } from '../src/parser/recipe/vanilla/shaped'
import createTestAcceptor from './mock/TestAcceptor'
import createTestLogger from './mock/TestLogger'
import createTestResolver from './mock/TestResolver'

const logger = createTestLogger()
const loader = new PackLoader(logger)

beforeAll(async () => {
   const resolver = createTestResolver()
   await loader.loadFrom(resolver)
}, 10_000)

afterEach(() => {
   loader.clear()
})

it('has no unknown recipe loaders', () => {
   expect(logger.warn).not.toHaveBeenCalled()
   expect(logger.error).not.toHaveBeenCalled()
})

describe('replace ingredient replacement', () => {
   it('replaces ingredients', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.replaceIngredient('minecraft:redstone', {
         item: 'minecraft:emerald',
      })

      await loader.emit(acceptor)

      expect(acceptor.jsonAt('data/minecraft/recipe/piston.json')).toMatchSnapshot('modified piston recipe')
      expect(acceptor.jsonAt('data/minecraft/recipe/compass.json')).toMatchSnapshot('modified compass recipe')

      expect(acceptor.paths()).toMatchSnapshot('recipes including redstone as an ingredient')
   })

   it('replaces ingredients with additional input filter', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.replaceIngredient(
         'minecraft:redstone',
         {
            item: 'minecraft:emerald',
         },
         {
            input: '#minecraft:planks',
         }
      )

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(2)

      expect(acceptor.jsonAt('data/minecraft/recipe/piston.json')).toMatchSnapshot('modified piston recipe')
      expect(acceptor.jsonAt('data/minecraft/recipe/note_block.json')).toMatchSnapshot('modified note_block recipe')
      expect(acceptor.jsonAt('data/minecraft/recipe/compass.json')).toBeNull()
   })

   it('replaces ingredients in create recipes', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.replaceIngredient('#forge:raw_materials/zinc', {
         tag: '#forge:raw_materials/iron',
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(4)

      expect(acceptor.jsonAt('data/create/recipe/crafting/materials/raw_zinc_block.json')).toMatchSnapshot(
         'modified create:raw_zinc_block recipe'
      )
      expect(acceptor.jsonAt('data/create/recipe/crushing/raw_zinc.json')).toMatchSnapshot(
         'modified create:raw_zinc recipe'
      )
      expect(acceptor.jsonAt('data/create/recipe/blasting/zinc_ingot_from_raw_ore.json')).toMatchSnapshot(
         'modified create:zinc_ingot_from_raw_ore recipe'
      )
      expect(acceptor.jsonAt('data/create/recipe/smelting/zinc_ingot_from_raw_ore.json')).toMatchSnapshot(
         'modified create:zinc_ingot_from_raw_ore recipe'
      )
   })
})

describe('recipe removal', () => {
   it('removes recipes with id filter', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.removeRecipe({
         id: /minecraft:.*piston/,
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(2)

      expect(acceptor.jsonAt('data/minecraft/recipe/piston.json')).toMatchObject(RecipeEmitter.EMPTY_RECIPE)
      expect(acceptor.jsonAt('data/minecraft/recipe/sticky_piston.json')).toMatchObject(RecipeEmitter.EMPTY_RECIPE)
   })

   it('removes recipes with result filter', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.removeRecipe({
         output: 'minecraft:cooked_beef',
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(3)

      expect(acceptor.jsonAt('data/minecraft/recipe/cooked_beef.json')).toMatchObject(RecipeEmitter.EMPTY_RECIPE)
      expect(acceptor.jsonAt('data/minecraft/recipe/cooked_beef_from_smoking.json')).toMatchObject(
         RecipeEmitter.EMPTY_RECIPE
      )
      expect(acceptor.jsonAt('data/minecraft/recipe/cooked_beef_from_campfire_cooking.json')).toMatchObject(
         RecipeEmitter.EMPTY_RECIPE
      )
   })
})

it('creates custom recipes', async () => {
   const acceptor = createTestAcceptor()

   const recipe: ShapedRecipeDefinition = {
      type: 'minecraft:shaped',
      key: {
         A: {
            item: 'minecraft:diamond',
         },
         B: {
            tag: 'minecraft:iron_ores',
         },
      },
      result: {
         item: 'minecraft:command_block',
      },
      pattern: ['A ', ' B'],
   }

   loader.recipes.addRecipe('example:custom', recipe)

   await loader.emit(acceptor)

   expect(acceptor.jsonAt('data/example/recipe/custom.json')).toMatchObject(recipe)
})
