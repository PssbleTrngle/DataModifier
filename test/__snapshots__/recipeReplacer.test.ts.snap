// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaces ingredients 1`] = `
{
  "key": {
    "#": {
      "item": "minecraft:cobblestone",
    },
    "R": {
      "item": "minecraft:emerald",
    },
    "T": {
      "tag": "minecraft:planks",
    },
    "X": {
      "item": "minecraft:iron_ingot",
    },
  },
  "pattern": [
    "TTT",
    "#X#",
    "#R#",
  ],
  "result": {
    "item": "minecraft:piston",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replaces ingredients 2`] = `
{
  "key": {
    "#": {
      "item": "minecraft:iron_ingot",
    },
    "X": {
      "item": "minecraft:emerald",
    },
  },
  "pattern": [
    " # ",
    "#X#",
    " # ",
  ],
  "result": {
    "item": "minecraft:compass",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replaces ingredients in create recipes 1`] = `
{
  "key": {
    "C": {
      "tag": "#forge:raw_materials/iron",
    },
  },
  "pattern": [
    "CCC",
    "CCC",
    "CCC",
  ],
  "result": {
    "item": "create:raw_zinc_block",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replaces ingredients in create recipes 2`] = `
{
  "ingredients": [
    {
      "tag": "#forge:raw_materials/iron",
    },
  ],
  "processingTime": 400,
  "results": [
    {
      "item": "create:crushed_raw_zinc",
    },
    {
      "chance": 0.75,
      "item": "create:experience_nugget",
    },
  ],
  "type": "create:crushing",
}
`;

exports[`replaces ingredients in create recipes 3`] = `
{
  "cookingtime": 100,
  "experience": 0.7,
  "ingredient": {
    "tag": "#forge:raw_materials/iron",
  },
  "result": "create:zinc_ingot",
  "type": "minecraft:blasting",
}
`;

exports[`replaces ingredients in create recipes 4`] = `
{
  "cookingtime": 200,
  "experience": 0.7,
  "ingredient": {
    "tag": "#forge:raw_materials/iron",
  },
  "result": "create:zinc_ingot",
  "type": "minecraft:smelting",
}
`;

exports[`replaces ingredients with additional input filter 1`] = `
{
  "key": {
    "#": {
      "item": "minecraft:cobblestone",
    },
    "R": {
      "item": "minecraft:emerald",
    },
    "T": {
      "tag": "minecraft:planks",
    },
    "X": {
      "item": "minecraft:iron_ingot",
    },
  },
  "pattern": [
    "TTT",
    "#X#",
    "#R#",
  ],
  "result": {
    "item": "minecraft:piston",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replaces ingredients with additional input filter 2`] = `
{
  "key": {
    "#": {
      "tag": "minecraft:planks",
    },
    "X": {
      "item": "minecraft:emerald",
    },
  },
  "pattern": [
    "###",
    "#X#",
    "###",
  ],
  "result": {
    "item": "minecraft:note_block",
  },
  "type": "minecraft:crafting_shaped",
}
`;
